File: apps/android/app/src/main/kotlin/com/hushos/app/data/Links.kt
package com.hushos.app.data

import com.hushos.core.ByteRange
import com.hushos.core.Content
import com.hushos.core.LinkContext
import com.hushos.core.MetadataContext
import com.hushos.core.NodeKeyContext
import com.hushos.core.OperatorKem
import com.hushos.core.OperatorKeys
import com.hushos.core.ReportContext
import com.hushos.core.VersionContext
import com.hushos.core.base64urlDecode
import com.hushos.core.base64urlEncode
import com.hushos.core.chunkDecrypt
import com.hushos.core.chunkRange
import com.hushos.core.linkOpen
import com.hushos.core.linkParse
import com.hushos.core.metadataOpen
import com.hushos.core.nodeOpen
import com.hushos.core.reportSealKey
import com.hushos.core.thumbnailDecrypt
import com.hushos.core.thumbnailRange
import com.hushos.core.versionOpen
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

/* A link as the owner sees it: what `GET /api/drive/nodes/:id/links` lists. */
data class LinkView(
    val id: String, val keyEpoch: ULong, val secretEnvelope: String?, val hasPassword: Boolean,
    val expiresAt: String?, val useCount: Int, val lastUsedAt: String?, val createdAt: String,
) {
    companion object {
        fun from(json: JSONObject) = LinkView(
            json.getString("id"), json.getLong("keyEpoch").toULong(),
            if (json.isNull("secretEnvelope")) null else json.getString("secretEnvelope"), json.getBoolean("hasPassword"),
            if (json.isNull("expiresAt")) null else json.getString("expiresAt"), json.optInt("useCount"),
            if (json.isNull("lastUsedAt")) null else json.getString("lastUsedAt"), json.getString("createdAt"),
        )
    }
}

/* What the server hands a visitor for a token: enough to open the root with the fragment secret. */
data class OpenedLink(
    val id: String, val workspaceId: String, val keyEpoch: ULong, val hasPassword: Boolean,
    val linkSalt: String, val linkEnvelope: String, val expiresAt: String?, val node: NodeView,
)

/* An operator as `GET /api/drive/reports/operators` lists them. */
data class ReportOperator(val userId: String, val encryptionPublicKey: String, val signingPublicKey: String, val kem: OperatorKem?)

object Reports {
    /* What can be reported, in the words the web uses. */
    val categories = listOf(
        "csam" to "Child sexual abuse material",
        "terrorism" to "Terrorist content",
        "ncii" to "Intimate images shared without consent",
        "malware" to "Malware",
        "copyright" to "Copyright infringement",
        "harassment" to "Harassment",
        "other" to "Something else",
    )

    /*
     * The report body: the node key sealed to every operator with the
     * report's context bound in, so the server, holding public keys only,
     * cannot open it, and an operator can open only this report.
     */
    fun body(operators: List<ReportOperator>, item: Opened, nodeKey: ByteArray, category: String, reason: String, email: String?, via: JSONObject): JSONObject {
        if (operators.isEmpty()) throw ApiError(503, "This HushOS has no operators to report to.")
        val reportId = UUID.randomUUID().toString()
        val keys = JSONArray()
        for (operator in operators) {
            val ctx = ReportContext(item.node.workspaceId, item.id, item.node.keyEpoch, reportId, operator.userId)
            val sealed = reportSealKey(ctx, nodeKey, OperatorKeys(operator.userId, operator.encryptionPublicKey, operator.signingPublicKey, operator.kem))
            keys.put(JSONObject().put("operatorUserId", operator.userId).put("keyEnvelope", base64urlEncode(sealed)))
        }
        val trimmed = email?.trim().orEmpty()
        return JSONObject().put("id", reportId).put("workspaceId", item.node.workspaceId).put("nodeId", item.id)
            .put("keyEpoch", item.node.keyEpoch.toLong()).put("category", category).put("reason", reason).put("via", via)
            .put("reporterEmail", if (trimmed.isEmpty()) JSONObject.NULL else trimmed).put("contentHash", JSONObject.NULL).put("keys", keys)
    }
}

private fun JSONArray.objects(): List<JSONObject> = (0 until length()).map { getJSONObject(it) }

/* The routes that need no session: what a visitor with a link, or anyone filing a report, talks to. */
class PublicApi(val origin: String) {
    private fun open(path: String, method: String): HttpURLConnection {
        val connection = URL("$origin/api/drive$path").openConnection() as HttpURLConnection
        connection.requestMethod = method
        connection.setRequestProperty("HushOS-Client", "android/2")
        connection.setRequestProperty("Origin", origin)
        connection.connectTimeout = 30_000
        connection.readTimeout = 60_000
        return connection
    }

    private fun finish(connection: HttpURLConnection): JSONObject {
        val status: Int
        val text: String
        try {
            status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            text = stream?.bufferedReader()?.use { it.readText() } ?: ""
        } catch (error: java.io.IOException) {
            throw Unreachable()
        }
        if (status !in 200..299) {
            val message = runCatching { JSONObject(text).optString("message") }.getOrNull()
            if (status == 404) throw NotFound(if (message.isNullOrEmpty()) "This link no longer works." else message)
            throw ApiError(status, if (message.isNullOrEmpty()) "Request failed" else message)
        }
        return if (text.isEmpty()) JSONObject() else JSONObject(text)
    }

    private fun get(path: String): JSONObject = finish(open(path, "GET"))

    fun post(path: String, body: JSONObject): JSONObject {
        val connection = open(path, "POST")
        connection.doOutput = true
        connection.setRequestProperty("Content-Type", "application/json")
        connection.outputStream.use { it.write(body.toString().toByteArray()) }
        return finish(connection)
    }

    fun openLink(token: String): OpenedLink = get("/links/$token/open").let {
        val link = it.getJSONObject("link")
        OpenedLink(link.getString("id"), link.getString("workspaceId"), link.getLong("keyEpoch").toULong(), link.getBoolean("hasPassword"),
            link.getString("linkSalt"), link.getString("linkEnvelope"), if (link.isNull("expiresAt")) null else link.getString("expiresAt"),
            NodeView.from(it.getJSONObject("node")))
    }

    fun linkChildren(token: String, nodeId: String, after: String?): Listing =
        get("/links/$token/nodes/$nodeId/children" + (after?.let { "?after=" + java.net.URLEncoder.encode(it, "UTF-8") } ?: "")).let {
            Listing(NodeView.from(it.getJSONObject("folder")), it.getJSONArray("ancestors").objects().map(NodeView::from),
                it.getJSONArray("children").objects().map(NodeView::from), if (it.isNull("nextCursor")) null else it.getString("nextCursor"))
        }

    fun linkVersionUrl(token: String, versionId: String): String = get("/links/$token/versions/$versionId/url").getString("url")

    fun linkThumbnailUrl(token: String, versionId: String): String? {
        val urls = get("/links/$token/thumbnails?versions=$versionId").getJSONArray("urls")
        return if (urls.length() > 0) urls.getJSONObject(0).getString("url") else null
    }

    fun operators(): List<ReportOperator> = get("/reports/operators").getJSONArray("operators").objects().map {
        ReportOperator(it.getString("userId"), it.getString("encryptionPublicKey"), it.getString("signingPublicKey"),
            it.optJSONObject("kem")?.let { kem -> OperatorKem(kem.getString("publicKey"), kem.getString("signature")) })
    }

    fun range(url: String, range: ByteRange): ByteArray {
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.setRequestProperty("Range", "bytes=${range.start}-${range.end}")
        connection.connectTimeout = 30_000
        connection.readTimeout = 120_000
        val status = try { connection.responseCode } catch (error: java.io.IOException) { throw Unreachable() }
        if (status != 206 && status != 200) throw ApiError(status, "Range request failed")
        return connection.inputStream.use { it.readBytes() }
    }
}

/*
 * A visitor's view through one link: the root opened with the fragment secret
 * (and the password when the link has one), every node below it under the
 * usual node keys, and content by the link's own download routes.
 */
class LinkVault(origin: String, url: String) {
    val api = PublicApi(origin)
    val token: String
    private val secret: String
    private var link: OpenedLink? = null
    private val nodeKeys = HashMap<String, ByteArray>()
    private val opened = HashMap<String, Opened>()

    init {
        val parts = linkParse(url)
        token = parts.token
        secret = parts.secret
    }

    private fun load(): OpenedLink = link ?: api.openLink(token).also { link = it }

    fun needsPassword(): Boolean = load().hasPassword

    /* Opens the root: the node key from the secret and password, then its name. */
    fun open(password: String?): Opened {
        val fetched = load()
        val ctx = LinkContext(fetched.workspaceId, fetched.node.id, fetched.keyEpoch, fetched.id)
        val key = linkOpen(ctx, base64urlDecode(fetched.linkEnvelope), base64urlDecode(secret), password, base64urlDecode(fetched.linkSalt))
        val node = fetched.node
        val metadata = metadataOpen(MetadataContext(node.workspaceId, node.id, node.metadataVersion), key, base64urlDecode(node.metadataEnvelope))
        nodeKeys[node.id] = key
        return Opened(node, metadata).also { opened[node.id] = it }
    }

    private fun open(node: NodeView): Opened {
        opened[node.id]?.let { return it }
        val parentId = node.parentId ?: throw ApiError(500, "This item is outside the link.")
        val parentKey = nodeKeys[parentId] ?: throw ApiError(500, "This item is outside the link.")
        val key = nodeOpen(NodeKeyContext(node.workspaceId, node.id, parentId, node.parentKeyEpoch, node.keyEpoch), parentKey, base64urlDecode(node.keyEnvelope))
        val metadata = metadataOpen(MetadataContext(node.workspaceId, node.id, node.metadataVersion), key, base64urlDecode(node.metadataEnvelope))
        nodeKeys[node.id] = key
        return Opened(node, metadata).also { opened[node.id] = it }
    }

    fun children(folderId: String): List<Opened> {
        var after: String? = null
        val all = ArrayList<Opened>()
        do {
            val page = api.linkChildren(token, folderId, after)
            page.children.forEach { all.add(open(it)) }
            after = page.nextCursor
        } while (after != null)
        return all.sortedWith(compareBy<Opened> { !it.isFolder }.thenBy(String.CASE_INSENSITIVE_ORDER) { it.name })
    }

    private fun openVersion(item: Opened): OpenedVersion {
        val key = nodeKeys[item.id] ?: throw ApiError(500, "Node not opened")
        val current = item.node.currentVersion ?: throw NotFound("This file has no content yet.")
        val v = versionOpen(VersionContext(item.node.workspaceId, item.id, current.id, current.objectId), current.contentSuite, key,
            base64urlDecode(current.contentKeyEnvelope), current.plaintextSize?.toULongOrNull())
        return OpenedVersion(Content(item.node.workspaceId, current.objectId, current.contentSuite, v.key, base64urlDecode(current.contentNonce), v.plaintextSize, v.thumbnailBytes), current.id)
    }

    fun download(item: Opened, destination: File, progress: (Float) -> Unit = {}) {
        val v = openVersion(item)
        var url = api.linkVersionUrl(token, v.versionId)
        // Resumable like the drive's own downloads: a dropped chunk is fetched again, a `.part` is continued.
        Resumable.download(destination, v.layout.chunkCount, v.layout.chunkBytes.toLong(), progress = progress,
            onExpired = { url = api.linkVersionUrl(token, v.versionId) },
        ) { index -> chunkDecrypt(v.content, index, api.range(url, chunkRange(v.content.plaintextSize, index))) }
    }

    fun thumbnail(item: Opened): ByteArray? {
        if (!item.hasThumbnail) return null
        val v = openVersion(item)
        if (v.content.thumbnailBytes == 0u) return null
        val url = api.linkThumbnailUrl(token, v.versionId) ?: return null
        return thumbnailDecrypt(v.content, api.range(url, thumbnailRange(v.content.plaintextSize, v.content.thumbnailBytes)))
    }

    /* Files a report on the linked item, sealing its key to every operator; anonymous. */
    fun report(item: Opened, category: String, reason: String, email: String?): Boolean {
        val nodeKey = nodeKeys[item.id] ?: throw ApiError(500, "Node not opened")
        val body = Reports.body(api.operators(), item, nodeKey, category, reason, email, JSONObject().put("link", token))
        return api.post("/reports", body).optBoolean("duplicate")
    }
}

File: apps/ios/Packages/HushOSKit/Sources/HushOSKit/Links.swift
import Foundation
import HushOSCore

/* A link as the owner sees it: what `GET /api/drive/nodes/:id/links` lists. */
public struct LinkView: Codable, Sendable, Identifiable, Hashable {
    public let id: String
    public let keyEpoch: UInt64
    public let secretEnvelope: String?
    public let hasPassword: Bool
    public let expiresAt: String?
    public let useCount: Int
    public let lastUsedAt: String?
    public let createdAt: String
}

struct LinksResponse: Codable, Sendable { let links: [LinkView] }
struct LinkResponse: Codable, Sendable { let link: LinkView }

/* What the server hands a visitor for a token: enough to open the root with the fragment secret. */
public struct OpenedLink: Codable, Sendable {
    public struct Link: Codable, Sendable {
        public let id: String
        public let workspaceId: String
        public let keyEpoch: UInt64
        public let hasPassword: Bool
        public let linkSalt: String
        public let linkEnvelope: String
        public let expiresAt: String?
    }
    public let link: Link
    public let node: NodeView
}

extension DriveAPI {
    public func links(of nodeId: String, workspaceId: String) async throws -> [LinkView] {
        try await performPublic(try requestPublic("/nodes/\(nodeId)/links", query: [URLQueryItem(name: "workspaceId", value: workspaceId)]), as: LinksResponse.self).links
    }

    public func createLink(nodeId: String, body: [String: Any]) async throws -> LinkView {
        try await performPublic(try mutationPublic("/nodes/\(nodeId)/links", body: body), as: LinkResponse.self).link
    }

    public func revokeLink(_ linkId: String, workspaceId: String) async throws {
        struct Revoked: Decodable { let revoked: Bool }
        _ = try await performPublic(try mutationPublic("/links/\(linkId)", method: "DELETE", body: ["workspaceId": workspaceId]), as: Revoked.self)
    }
}

/*
 * The routes that need no session: what a visitor with a link, or anyone
 * filing a report, talks to. Same client header, no cookie.
 */
public final class PublicAPI: Sendable {
    public let origin: String
    private let http = URLSession(configuration: .ephemeral)

    public init(origin: String) { self.origin = origin }

    private func request(_ path: String, query: [URLQueryItem] = []) throws -> URLRequest {
        guard var components = URLComponents(string: origin + "/api/drive" + path) else { throw DriveAPIError.server(500, "Bad API path") }
        if !query.isEmpty { components.queryItems = query }
        guard let url = components.url else { throw DriveAPIError.server(500, "Bad API path") }
        var request = URLRequest(url: url)
        request.httpShouldHandleCookies = false
        request.setValue("ios/2", forHTTPHeaderField: "HushOS-Client")
        request.setValue(origin, forHTTPHeaderField: "Origin")
        request.timeoutInterval = 30
        return request
    }

    private func perform<T: Decodable>(_ request: URLRequest, as type: T.Type) async throws -> T {
        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await http.data(for: request)
        } catch {
            throw DriveAPIError.transport(error.localizedDescription)
        }
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200 ..< 300).contains(status) else {
            let message = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["message"] as? String
            if status == 404 { throw DriveAPIError.notFound(message ?? "This link no longer works.") }
            throw DriveAPIError.server(status, message ?? "Request failed")
        }
        return try JSONDecoder().decode(type, from: data)
    }

    func post<T: Decodable>(_ path: String, body: [String: Any], as type: T.Type) async throws -> T {
        var request = try request(path)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        return try await perform(request, as: type)
    }

    public func openLink(token: String) async throws -> OpenedLink {
        try await perform(try request("/links/\(token)/open"), as: OpenedLink.self)
    }

    public func linkChildren(token: String, nodeId: String, after: String?) async throws -> Listing {
        let query = after.map { [URLQueryItem(name: "after", value: $0)] } ?? []
        return try await perform(try request("/links/\(token)/nodes/\(nodeId)/children", query: query), as: Listing.self)
    }

    public func linkVersionURL(token: String, versionId: String) async throws -> VersionURL {
        try await perform(try request("/links/\(token)/versions/\(versionId)/url"), as: VersionURL.self)
    }

    public func linkThumbnails(token: String, versionIds: [String]) async throws -> ThumbnailUrls {
        try await perform(try request("/links/\(token)/thumbnails", query: [URLQueryItem(name: "versions", value: versionIds.joined(separator: ","))]), as: ThumbnailUrls.self)
    }

    public func range(_ url: URL, _ range: ByteRange) async throws -> Data {
        var request = URLRequest(url: url)
        request.setValue("bytes=\(range.start)-\(range.end)", forHTTPHeaderField: "Range")
        request.timeoutInterval = 120
        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await http.data(for: request)
        } catch {
            throw DriveAPIError.transport(error.localizedDescription)
        }
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard status == 206 || status == 200 else { throw DriveAPIError.server(status, "Range request failed") }
        return data
    }
}

extension Vault {
    /* The links the owner made for this item. */
    public func links(for item: Opened) async throws -> [LinkView] {
        try await api.links(of: item.id, workspaceId: item.node.workspaceId)
    }

    /* Mints a link for an item this account holds the key of, and returns the URL to hand out. */
    public func createLink(for item: Opened, password: String?, expiresAt: Date?) async throws -> (link: LinkView, url: URL) {
        guard let nodeKey = nodeKeys[item.id] else { throw DriveAPIError.server(500, "Node not opened") }
        let ctx = LinkContext(workspaceId: item.node.workspaceId, nodeId: item.id, keyEpoch: item.node.keyEpoch, linkId: UUID().uuidString.lowercased())
        let created = try linkCreate(ctx: ctx, nodeKey: nodeKey, password: password)
        var body: [String: Any] = [
            "workspaceId": item.node.workspaceId, "linkId": ctx.linkId, "token": created.token, "keyEpoch": item.node.keyEpoch,
            "linkEnvelope": created.linkEnvelope, "linkSalt": created.linkSalt, "secretEnvelope": created.secretEnvelope,
            "hasPassword": created.hasPassword, "expiresAt": NSNull(),
        ]
        if let expiresAt { body["expiresAt"] = ISO8601DateFormatter().string(from: expiresAt) }
        let link = try await api.createLink(nodeId: item.id, body: body)
        guard let url = URL(string: try linkUrl(origin: session.origin, token: created.token, secret: created.secret)) else {
            throw DriveAPIError.server(500, "Bad link URL")
        }
        return (link, url)
    }

    /* The URL of an existing link, from the owner's sealed copy of its secret. */
    public func linkURL(_ link: LinkView, for item: Opened) throws -> URL? {
        guard let nodeKey = nodeKeys[item.id], let sealed = link.secretEnvelope else { return nil }
        let ctx = LinkContext(workspaceId: item.node.workspaceId, nodeId: item.id, keyEpoch: link.keyEpoch, linkId: link.id)
        let owned = try linkSecretOpen(ctx: ctx, nodeKey: nodeKey, envelope: try base64urlDecode(value: sealed))
        return URL(string: try linkUrl(origin: session.origin, token: base64urlEncode(bytes: owned.token), secret: base64urlEncode(bytes: owned.secret)))
    }

    public func revokeLink(_ link: LinkView, for item: Opened) async throws {
        try await api.revokeLink(link.id, workspaceId: item.node.workspaceId)
    }
}

/*
 * A visitor's view through one link: the root opened with the fragment secret
 * (and the password when the link has one), every node below it under the
 * usual node keys, and content by the link's own download routes.
 */
public actor LinkVault {
    public let api: PublicAPI
    public let token: String
    private let secret: String
    private var link: OpenedLink?
    private var nodeKeys: [String: Data] = [:]
    private var opened: [String: Opened] = [:]

    public init(origin: String, url: String) throws {
        let parts = try linkParse(url: url)
        api = PublicAPI(origin: origin)
        token = parts.token
        secret = parts.secret
    }

    /* Whether the link wants a password; fetches the link once. */
    public func needsPassword() async throws -> Bool {
        try await load().link.hasPassword
    }

    private func load() async throws -> OpenedLink {
        if let link { return link }
        let fetched = try await api.openLink(token: token)
        link = fetched
        return fetched
    }

    public func rootId() async throws -> String { try await load().node.id }

    /* Opens the root: the node key from the secret and password, then its name. */
    public func open(password: String?) async throws -> Opened {
        let fetched = try await load()
        let ctx = LinkContext(workspaceId: fetched.link.workspaceId, nodeId: fetched.node.id, keyEpoch: fetched.link.keyEpoch, linkId: fetched.link.id)
        let key = try linkOpen(
            ctx: ctx, envelope: try base64urlDecode(value: fetched.link.linkEnvelope), secret: try base64urlDecode(value: secret),
            password: password, salt: try base64urlDecode(value: fetched.link.linkSalt)
        )
        let node = fetched.node
        let metadata = try metadataOpen(
            ctx: MetadataContext(workspaceId: node.workspaceId, nodeId: node.id, metadataVersion: node.metadataVersion),
            nodeKey: key, envelope: try base64urlDecode(value: node.metadataEnvelope)
        )
        nodeKeys[node.id] = key
        let root = Opened(node: node, metadata: metadata)
        opened[node.id] = root
        return root
    }

    private func open(_ node: NodeView) throws -> Opened {
        if let done = opened[node.id] { return done }
        guard let parentId = node.parentId, let parentKey = nodeKeys[parentId] else {
            throw DriveAPIError.server(500, "This item is outside the link.")
        }
        let ctx = NodeKeyContext(workspaceId: node.workspaceId, nodeId: node.id, parentId: parentId, parentKeyEpoch: node.parentKeyEpoch, keyEpoch: node.keyEpoch)
        let key = try nodeOpen(ctx: ctx, parentKey: parentKey, keyEnvelope: try base64urlDecode(value: node.keyEnvelope))
        let metadata = try metadataOpen(
            ctx: MetadataContext(workspaceId: node.workspaceId, nodeId: node.id, metadataVersion: node.metadataVersion),
            nodeKey: key, envelope: try base64urlDecode(value: node.metadataEnvelope)
        )
        nodeKeys[node.id] = key
        let result = Opened(node: node, metadata: metadata)
        opened[node.id] = result
        return result
    }

    /* A folder's children, opened and sorted folders first. */
    public func children(of folderId: String) async throws -> [Opened] {
        var after: String? = nil
        var all: [Opened] = []
        repeat {
            let page = try await api.linkChildren(token: token, nodeId: folderId, after: after)
            for child in page.children { all.append(try open(child)) }
            after = page.nextCursor
        } while after != nil
        return all.sorted(by: Opened.byName)
    }

    private func openVersion(_ item: Opened) throws -> OpenedVersion {
        guard let nodeKey = nodeKeys[item.id] else { throw DriveAPIError.server(500, "Node not opened") }
        guard let current = item.node.currentVersion else { throw DriveAPIError.notFound("This file has no content yet.") }
        let opened = try HushOSCore.versionOpen(
            ctx: VersionContext(workspaceId: item.node.workspaceId, nodeId: item.id, versionId: current.id, objectId: current.objectId),
            suite: current.contentSuite, nodeKey: nodeKey, envelope: try base64urlDecode(value: current.contentKeyEnvelope),
            rowSize: current.plaintextSize.flatMap { UInt64($0) }
        )
        let content = Content(
            workspaceId: item.node.workspaceId, objectId: current.objectId, suite: current.contentSuite, key: opened.key,
            nonce: try base64urlDecode(value: current.contentNonce), plaintextSize: opened.plaintextSize, thumbnailBytes: opened.thumbnailBytes
        )
        return OpenedVersion(content: content, versionId: current.id)
    }

    public func download(_ item: Opened, to destination: URL, progress: @Sendable (Double) -> Void = { _ in }) async throws {
        let opened = try openVersion(item)
        let url = try await api.linkVersionURL(token: token, versionId: opened.versionId)
        guard var objectURL = URL(string: url.url) else { throw DriveAPIError.server(500, "Bad object URL") }
        // Resumable: a dropped chunk is fetched again, an expired address renewed, and a `.part`
        // an earlier try left is continued from its last whole chunk; it moves into place whole.
        try await Resumable.download(
            to: destination, count: opened.layout.chunkCount, chunkBytes: opened.layout.chunkBytes,
            onExpired: { if let fresh = URL(string: (try await api.linkVersionURL(token: token, versionId: opened.versionId)).url) { objectURL = fresh } },
            progress: progress
        ) { index in
            let ciphertext = try await api.range(objectURL, chunkRange(plaintextSize: opened.content.plaintextSize, index: index))
            return try chunkDecrypt(content: opened.content, index: index, ciphertext: ciphertext)
        }
    }

    public func thumbnail(_ item: Opened) async throws -> Data? {
        guard item.hasThumbnail, let version = item.node.currentVersion else { return nil }
        let opened = try openVersion(item)
        guard opened.content.thumbnailBytes > 0 else { return nil }
        let urls = try await api.linkThumbnails(token: token, versionIds: [version.id])
        guard let entry = urls.urls.first, let url = URL(string: entry.url) else { return nil }
        let ciphertext = try await api.range(url, thumbnailRange(plaintextSize: opened.content.plaintextSize, thumbnailBytes: opened.content.thumbnailBytes))
        return try thumbnailDecrypt(content: opened.content, ciphertext: ciphertext)
    }

    /* Files a report on the linked item, sealing its key to every operator; anonymous. */
    public func report(_ item: Opened, category: String, reason: String, email: String?) async throws -> Bool {
        guard let nodeKey = nodeKeys[item.id] else { throw DriveAPIError.server(500, "Node not opened") }
        let body = try await Reports.body(api: api, item: item, nodeKey: nodeKey, category: category, reason: reason, email: email, via: ["link": token])
        return try await api.post("/reports", body: body, as: ReportFiled.self).duplicate
    }
}
